Author: Luca Boccassi <luca.boccassi@gmail.com>
Description: Fix kernel module load on 4.11 and greater
 From kernel 4.11 and newer (11bac80004) the vm_area_struct member was removed
 from the fault callback signature, as it is already contained in struct
 vm_fault.
--- a/uvm/nvidia_uvm_lite.c
+++ b/uvm/nvidia_uvm_lite.c
@@ -30,6 +30,8 @@
 #include "nvidia_uvm_lite_counters.h"
 #include "ctrl2080mc.h"
 
+#include <linux/version.h>
+
 //
 // nvidia_uvm_lite.c
 // This file contains code that is specific to the UVM-Lite mode of operation.
@@ -689,9 +691,16 @@ static UvmGpuMigrationTracking * _get_mi
 //         return SIGBUS.
 //      3. Otherwise, map in a page from the cache, and allow access.
 //
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
+int _fault_common(struct vm_fault *vmf, unsigned long vaddr,
+#else
 int _fault_common(struct vm_area_struct *vma, unsigned long vaddr,
+#endif
                   struct page **ppage, unsigned vmfFlags)
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
+    struct vm_area_struct *vma = vmf->vma;
+#endif
     int retValue = VM_FAULT_SIGBUS;
     UvmCommitRecord *pRecord;
     DriverPrivate *pPriv;
@@ -818,7 +827,11 @@ done:
 }
 
 #if defined(NV_VM_OPERATIONS_STRUCT_HAS_FAULT)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
+int _fault(struct vm_fault *vmf)
+#else
 int _fault(struct vm_area_struct *vma, struct vm_fault *vmf)
+#endif
 {
 #if defined(NV_VM_FAULT_HAS_ADDRESS)
     unsigned long vaddr = vmf->address;
@@ -828,7 +841,11 @@ int _fault(struct vm_area_struct *vma, s
     struct page *page = NULL;
     int retval;
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
+    retval = _fault_common(vmf, vaddr, &page, vmf->flags);
+#else
     retval = _fault_common(vma, vaddr, &page, vmf->flags);
+#endif
 
     vmf->page = page;
 
@@ -866,7 +883,11 @@ static struct vm_operations_struct uvmli
 // it's dealing with anonymous mapping (see handle_pte_fault).
 //
 #if defined(NV_VM_OPERATIONS_STRUCT_HAS_FAULT)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
+int _sigbus_fault(struct vm_fault *vmf)
+#else
 int _sigbus_fault(struct vm_area_struct *vma, struct vm_fault *vmf)
+#endif
 {
     vmf->page = NULL;
     return VM_FAULT_SIGBUS;
